Fix Part of #4938: Create Admin  Onboarding Intro(2/12)

adhiamboperes opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes part of #4938  

This PR creates the introduction screen, where the user is informed about what it means to be an admin.  

In https://github.com/oppia/oppia-android/pull/5378, the "I am a parent/Teacher" option previously routed to the profiles list screen, since admin onboarding had not yet been implemented. In this PR, the option leads to the admin creating their profile nickname, followed by viewing the newly introduced introduction screen. The steps are ordered this way for consistency with the learner onboarding flow.  

In https://github.com/oppia/oppia-android/pull/5387, supervisor profile onboarding was marked as started on the profile type screen and ended on the home screen. This workaround has now been updated so that supervisor onboarding starts on the introduction screen (similar to the learner flow). This means that if the user abandons onboarding at this point, they will resume the flow from the introduction screen when they reopen the app. Admin onboarding is marked as completed on the profile chooser screen.  

This PR also ensures that the onboarding event logs are refactored accordingly, and tests have been updated to verify the refactor.  

Additionally, `testFragment_profileTypeArgumentMissing_showsUnknownProfileTypeError()` has been removed from `CreateProfileFragmentTest.kt` because this condition will never be hit, due to a new check that verifies `ProfileType` is not null when a `CreateProfileActivity` intent is created.  

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-A11y-Guide))
- For PRs introducing new UI elements or color changes, both light and dark mode screenshots must be included
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

|Landscape|Portrait|
|---|---|
|![Screenshot_1751638663](https://github.com/user-attachments/assets/52f767a3-d5f5-4207-b813-45bb34c400e4)|![Screenshot_1751637334](https://github.com/user-attachments/assets/6783dd8c-97bd-48f5-98ae-d7fc4476fd32)|
|![Screenshot_1751640466](https://github.com/user-attachments/assets/d5c21c41-e73a-4254-8647-894e03fdc0a1)|![Screenshot_1751639602](https://github.com/user-attachments/assets/3625d48f-2108-4790-8d46-231a8f9116f4)|
| https://github.com/user-attachments/assets/b1dd5108-f439-4817-bdab-40f676f86d75 | https://github.com/user-attachments/assets/c4cf5c05-722d-4d5a-baca-15752cc7b3ee |
| [tablet landscape](https://github.com/user-attachments/assets/9fac2949-dd88-4052-bd96-fd41889beadc)  | [tablet_portrait](https://github.com/user-attachments/assets/25d03a33-3a77-4ffa-830f-44f0a09fe6cc) |
| [event_logs.webm](https://github.com/user-attachments/assets/46ba25e8-379f-443f-ade3-440e1647be68) | <img width="1600" height="2560" alt="Screenshot_1758651952" src="https://github.com/user-attachments/assets/bb0a3127-0e49-4d4d-97d5-04811017a1d3" />|

adhiamboperes committed:
New layout for auth

adhiamboperes committed:
Merge remote-tracking branch 'upstream/develop' into redesign-pin-input

# Conflicts:
#	app/src/main/res/values/styles.xml

adhiamboperes committed:
Add compose livedata dependency

adhiamboperes committed:
Create color and string resources

adhiamboperes committed:
Create xml layouts

adhiamboperes committed:
Create ProfileLoginActivity

adhiamboperes committed:
Create ProfileLoginFragment

adhiamboperes committed:
Refactor login intent to new ProfileLoginActivity

adhiamboperes committed:
Add todos to track cleanup

adhiamboperes committed:
Cleanup unused items

adhiamboperes committed:
Add admin reset pin flow

adhiamboperes committed:
Add route to non-admin reset pin flow

adhiamboperes committed:
Add non-admin pin reset flow

adhiamboperes committed:
wip commit

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into redesign-pin-input

adhiamboperes committed:
Create AdminIntroActivity

adhiamboperes committed:
Create AdminIntroFragment

adhiamboperes committed:
Correct ktlint errors

adhiamboperes committed:
Add navigation to the screen

adhiamboperes committed:
Create the screen components

adhiamboperes committed:
Add navigation to the next screen

adhiamboperes committed:
Add activity tests

adhiamboperes committed:
Add fragment tests

adhiamboperes committed:
Break down large composables

adhiamboperes committed:
Add tests for login flow

adhiamboperes committed:
Move dialogs to own files

adhiamboperes committed:
Add tests for dialog flows: wip for non-admin

adhiamboperes committed:
Add todo to ensure an admin profile is created when all profile data is reset

adhiamboperes committed:
Fix lint issues

adhiamboperes committed:
Fix static analysis checks failures

adhiamboperes committed:
Fix ProfileChooserFragmentTest failures

adhiamboperes committed:
Finish dialog tests

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into redesign-pin-input

adhiamboperes committed:
Format dagger modules

adhiamboperes committed:
Fix missing dagger module

adhiamboperes committed:
Merge remote-tracking branch 'upstream/redesign-pin-input' into create-admin-intro

# Conflicts:
#	app/src/main/java/org/oppia/android/app/profile/ProfileChooserActivityPresenter.kt
#	app/src/main/java/org/oppia/android/app/profile/ProfileLoginFragmentPresenter.kt

adhiamboperes committed:
Ensure navigation is correct

Moved start onboarding to intro activity

Temporarily navigate to profile chooser until create profile nickname is completed

Ensure all activities in the backstack are removed

adhiamboperes committed:
Cleanup navigation and add tests for all changes

adhiamboperes committed:
Fix UI and failing tests

adhiamboperes committed:
fix trim spannable bug

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into develop

adhiamboperes committed:
Merge branch 'develop' into redesign-pin-input

adhiamboperes committed:
Merge branch 'redesign-pin-input' into create-admin-intro

adhiamboperes committed:
Merge branch 'develop' into redesign-pin-input

# Conflicts:
#	app/src/main/java/org/oppia/android/app/fragment/FragmentComponentImpl.kt
#	app/src/main/res/values-night/color_palette.xml
#	app/src/main/res/values/color_palette.xml
#	app/src/main/res/values/strings.xml
#	third_party/maven_install.json

adhiamboperes committed:
Generate a new maven_install due to conflicts in the old one.

adhiamboperes committed:
regenerate maven_install.json

adhiamboperes committed:
address reviewer comment.

adhiamboperes committed:
Revert unrelated changed

adhiamboperes committed:
fix tests

adhiamboperes committed:
Merge branch 'develop' into redesign-pin-input

# Conflicts:
#	app/src/main/res/values/color_palette.xml
#	third_party/maven_install.json

adhiamboperes committed:
Merge branch 'redesign-pin-input' into create-admin-intro

# Conflicts:
#	app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt
#	app/src/main/res/values-land/dimens.xml
#	app/src/main/res/values-sw600dp-land/dimens.xml
#	app/src/main/res/values-sw600dp-port/dimens.xml
#	model/src/main/proto/arguments.proto

adhiamboperes committed:
Merge branch 'develop' into redesign-pin-input

adhiamboperes committed:
fix conflicts

adhiamboperes committed:
Repin dependencies

adhiamboperes committed:
Merge branch 'redesign-pin-input' of github.com:oppia/oppia-android into redesign-pin-input

adhiamboperes committed:
Merge branch 'redesign-pin-input' into create-admin-intro

adhiamboperes committed:
Fix ktlint issue

adhiamboperes committed:
Fix file formatting

adhiamboperes committed:
Merge branch 'develop' into redesign-pin-input

adhiamboperes committed:
Exempt files from coverage.

adhiamboperes committed:
Address reviewer comments

adhiamboperes committed:
Fix coverage failures

adhiamboperes committed:
Merge branch 'redesign-pin-input' into create-admin-intro

adhiamboperes committed:
Minor cleanup

adhiamboperes committed:
Fix Kdoc validity check

adhiamboperes committed:
Fix navigation bug

adhiamboperes committed:
Fix wavy background

adhiamboperes committed:
Fix lint checks

adhiamboperes commented:
Hi @deonwaju, @subhajitxyz, please do a review pass. Thanks.

subhajitxyz commented on app/src/main/res/layout/admin_intro_activity.xml:
```diff
@@ -0,0 +1,6 @@
+    android:id="@+id/admin_intro_fragment_placeholder"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:orientation="vertical" />
```
`orientation` isn’t applicable to FrameLayout, so it can be safely removed.

adhiamboperes replied:
Removed, thanks!

subhajitxyz commented on app/src/main/java/org/oppia/android/app/onboarding/AdminIntroFragmentPresenter.kt:
```diff
@@ -0,0 +1,310 @@
+/** Test tag for the otter image. */
+const val OTTER_TEST_TAG = "TEST_TAG.otter"
+
+/** Test tag for the otter image. */
```
I think , KDoc comment needs to updated.

adhiamboperes replied:
Thanks for pointing that out. I however deleted the variable and the corresponding KDoc in latest changes.

subhajitxyz commented on app/src/main/java/org/oppia/android/app/onboarding/AdminIntroFragmentPresenter.kt:
```diff
@@ -0,0 +1,310 @@
+/** Test tag for the otter image. */
+const val BACK_BUTTON_TEST_TAG = "TEST_TAG.back_button"
+
+/** Test tag for the otter image. */
```
I think the KDoc comment needs to be updated.

adhiamboperes replied:
I deleted the variable and the corresponding KDoc in latest changes.

subhajitxyz commented on app/src/main/java/org/oppia/android/app/onboarding/WavyBackgroundView.kt:
```diff
@@ -0,0 +1,54 @@
+/** Adds a bezier curve background to a view. */
+@Composable
+fun WavyBackgroundView(backgroundColorResId: Int) {
+  val whiteColor = colorResource(backgroundColorResId)
```
```suggestion
  val waveColor = colorResource(backgroundColorResId)
```

adhiamboperes replied:
Updated.

subhajitxyz commented on app/src/main/java/org/oppia/android/app/splash/SplashActivityPresenter.kt:
```diff
@@ -438,27 +443,40 @@ class SplashActivityPresenter @Inject constructor(
+          ProfileType.SUPERVISOR -> launchProfileChooserScreen()
+          else -> {}
         }
-        else -> launchOnboardingActivity()
```
Is it possible that profile != null && profile.startedProfileOnboarding == false ??

adhiamboperes replied:
Great question! This condition would not be hit because [this diif](https://github.com/oppia/oppia-android/pull/5839/files#diff-186c157f395f5c3d6503d20483fc97e4229918428c7224a70dfe4a847cc80435R449-R451) already handles the `profile==null` case by launching the Onboarding Screen in such cases.

BenHenning replied:
Perhaps we should combine them together so that the else cases are impossible? I do worry a bit about the else cases since they would make the user stuck, I think.

adhiamboperes replied:
I have made some changes. PTAL. I wonder if this scenario would benefit from a sealed class since the logic os getting quite complex.

BenHenning replied:
I've been reading a bit of Kotlin 2 changes and it looks like this would probably be a really good candidate for guarded conditions, but alas. Maybe we can update to Kotlin 2.x sometime next year.

Sealed classes may well be a good alternative that we could use now, but I think what you have now is completely reasonable.

Feel free to resolve this--I'm happy with it.

subhajitxyz commented on app/src/sharedTest/java/org/oppia/android/app/onboarding/AdminIntroActivityTest.kt:
```diff
@@ -0,0 +1,228 @@
+  fun testActivity_hasCorrectActivityLabel() {
+    launch<AdminIntroActivity>(createAdminIntroActivityIntent())
+      .use { scenario ->
+        scenario?.onActivity { activity ->
```
Just curious—why is scenario treated as nullable here?

adhiamboperes replied:
it should not have been. I have removed the the unnecessary safe call.

subhajitxyz commented:
Thanks @adhiamboperes , the UI looks awesome! I’ve taken an initial pass and left some comments.

adhiamboperes committed:
Merge branch 'develop' into redesign-pin-input

# Conflicts:
#	model/src/main/proto/screens.proto
#	utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt

adhiamboperes committed:
Merge branch 'redesign-pin-input' into create-admin-intro

# Conflicts:
#	model/src/main/proto/screens.proto

adhiamboperes committed:
Merge branch 'develop' into redesign-pin-input

# Conflicts:
#	model/src/main/proto/screens.proto
#	utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt

adhiamboperes committed:
Merge branch 'redesign-pin-input' into create-admin-intro

# Conflicts:
#	model/src/main/proto/screens.proto

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into redesign-pin-input

adhiamboperes committed:
Merge branch 'develop' into redesign-pin-input

adhiamboperes committed:
Use use {} block for scenario

adhiamboperes committed:
Remove test tags

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into redesign-pin-input

adhiamboperes committed:
Merge branch 'redesign-pin-input' of github.com:oppia/oppia-android into redesign-pin-input

adhiamboperes committed:
Add lint exemption

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into redesign-pin-input

adhiamboperes committed:
Add comments explaining test tag usage

adhiamboperes committed:
fix kdoc formatting

adhiamboperes committed:
add comment

adhiamboperes committed:
resolve merge conflicts from develop

adhiamboperes committed:
Remove duplicate tests

adhiamboperes committed:
Merge remote-tracking branch 'upstream/develop' into redesign-pin-input

adhiamboperes commented on app/src/main/java/org/oppia/android/app/onboarding/CreateProfileFragmentPresenter.kt:
```diff
@@ -161,56 +161,55 @@ class CreateProfileFragmentPresenter @Inject constructor(
+      when (result) {
+        is AsyncResult.Success -> {
+          createProfileViewModel.hasErrorMessage.set(false)
+          // TODO(#4938): Add navigation to the PIN creation screen for ProfileType.SUPERVISOR
```
Remove TODO since the updated behaviour is that we show intro screen after profile creation to be consistent with the learner flow

adhiamboperes replied:
Done

adhiamboperes committed:
Merge branch 'redesign-pin-input' into create-admin-intro

# Conflicts:
#	app/src/main/java/org/oppia/android/app/onboarding/CreateProfileFragmentPresenter.kt

adhiamboperes commented on app/src/sharedTest/java/org/oppia/android/app/onboarding/AdminIntroFragmentTest.kt:
```diff
@@ -0,0 +1,323 @@
+
+    testCoroutineDispatchers.runCurrent()
+
+    intended(hasComponent(ProfileChooserActivity::class.java.name))
```
Add comment that this should fail when the pin setup screen has been implemented.

adhiamboperes replied:
Done

adhiamboperes commented on app/src/sharedTest/java/org/oppia/android/app/onboarding/AdminIntroFragmentTest.kt:
```diff
@@ -0,0 +1,323 @@
+    composeRule.onNodeWithText(context.getString(R.string.admin_intro_activity_learners_text))
+      .assertIsDisplayed()
+
+    composeRule.onNodeWithText(context.getString(R.string.onboarding_step_count_three))
```
Should be 4

adhiamboperes replied:
Done

adhiamboperes committed:
Fix indentation issue

adhiamboperes committed:
General cleanup.

adhiamboperes committed:
Refactor some intents.

adhiamboperes committed:
Ensure the newly created admin name is picked up by the profile chooser.

adhiamboperes committed:
Refactor complete onboarding logic and update tests

adhiamboperes committed:
Update AdminIntroFragmentTest tests

adhiamboperes committed:
Update CreateProfileFragmentTest tests

adhiamboperes commented on app/src/sharedTest/java/org/oppia/android/app/onboarding/CreateProfileFragmentTest.kt:
```diff
@@ -644,32 +706,6 @@ class CreateProfileFragmentTest {
   }
 
-  @Test
-  fun testFragment_profileTypeArgumentMissing_showsUnknownProfileTypeError() {
```
Document this removal

adhiamboperes replied:
Done.

adhiamboperes commented on app/src/main/java/org/oppia/android/app/classroom/ClassroomListFragmentPresenter.kt:
```diff
@@ -275,13 +275,27 @@ class ClassroomListFragmentPresenter @Inject constructor(
         val profile = result.value
         val profileType = profile.profileType
 
-        if (enableOnboardingFlowV2.value && !profile.completedProfileOnboarding) {
```
Document this refactor

adhiamboperes replied:
Done

adhiamboperes committed:
fix failing tests

adhiamboperes reviewed (commented):
Self-reviewed.

adhiamboperes committed:
address reviewer comments

adhiamboperes committed:
Update admin profile creation and profile choose launch

adhiamboperes committed:
Fix failing admin intro test and lint errors

adhiamboperes committed:
Fix failing OnboardingProfileTypeFragmentTest tests

adhiamboperes committed:
Merge branch 'develop' into create-admin-intro

# Conflicts:
#	app/src/main/AndroidManifest.xml
#	app/src/main/java/org/oppia/android/app/activity/ActivityComponentImpl.kt
#	app/src/main/java/org/oppia/android/app/fragment/FragmentComponentImpl.kt
#	app/src/main/java/org/oppia/android/app/profile/ProfileChooserActivityPresenter.kt
#	app/src/main/res/values/strings.xml
#	model/src/main/proto/screens.proto
#	utility/src/main/java/org/oppia/android/util/logging/EventBundleCreator.kt

adhiamboperes commented:
Thanks for your initial review @subhajitxyz! I addressed your comments. However, there have been a lot of refactors since your review, and I appreciate that you might not have time to review a second time due to being on a break. 


@BenHenning, PTAL.

BenHenning commented:
Apologies--will try to look at this early next week. Feel free to send me more chained PRs as well if that helps.

BenHenning commented on app/src/main/java/org/oppia/android/app/classroom/ClassroomListFragmentPresenter.kt:
```diff
@@ -275,13 +275,27 @@ class ClassroomListFragmentPresenter @Inject constructor(
+        if (enableOnboardingFlowV2.value) {
+          when {
+            profileType == ProfileType.SUPERVISOR && profile.numberOfLogins == 1 -> {
+              // Supervisors end  profile onboarding on the profiles list screen, but they do
```
```suggestion
              // Supervisors end profile onboarding on the profiles list screen, but they do
```

Small nit.

adhiamboperes replied:
Fixed

BenHenning commented on app/src/main/java/org/oppia/android/app/onboarding/WavyBackgroundView.kt:
```diff
@@ -0,0 +1,54 @@
+import androidx.compose.ui.unit.dp
+import kotlin.math.sin
+
+/** Adds a bezier curve background to a view. */
```
Slight nit maybe, but is this actually a bezier curve? It appears to be a sine wave, instead.

adhiamboperes replied:
Well, it is technically a sine wave. Whe I originally wrote this for Android Views I used cubic to create a bezier curve, but for some reason only sine would work with compose.

adhiamboperes replied:
Updated the kdoc.

BenHenning commented on app/src/main/res/layout/admin_intro_fragment.xml:
```diff
@@ -0,0 +1,7 @@
+<layout xmlns:android="http://schemas.android.com/apk/res/android">
+    <androidx.compose.ui.platform.ComposeView
```
For this & app/src/main/res/layout/admin_intro_activity.xml these should be two-space indented.

adhiamboperes replied:
Fixed.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/onboarding/AdminIntroFragmentTest.kt:
```diff
@@ -0,0 +1,359 @@
+  @Test
+  fun testIntroFragment_onLaunch_landscapeMode_allViewsAreCorrectlyDisplayed() {
+    launch(AdminIntroActivity::class.java).use {
+      onView(ViewMatchers.isRoot()).perform(orientationLandscape())
```
This test should start in landscape using a qualifier rather than rotating (since rotating isn't 100% reliable in Robolectric).

Also, shouldn't this verify that the step count isn't visible since that's orientation-specific logic?

adhiamboperes replied:
Done for both points.

BenHenning reviewed (changes requested):
Thanks @adhiamboperes! Took a full pass and had just a few comments--PTAL.

adhiamboperes committed:
Address reviewer comments

adhiamboperes committed:
Merge branch 'develop' into create-admin-intro

# Conflicts:
#	app/src/main/java/org/oppia/android/app/onboarding/CreateProfileActivity.kt

adhiamboperes committed:
Fix merge conflict

BenHenning reviewed (approved):
Thanks @adhiamboperes! This LGTM--no further changes needed from my perspective.

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into create-admin-intro

adhiamboperes committed:
Fix failing tests

adhiamboperes reviewed (commented):
Self-reviewed latest changes. Fixed one test.

adhiamboperes commented:
Synchronized branch and enabling auto-merge.
